library/alloc/tests/str.rs

```diff
@@ -2141,3 +2141,43 @@ fn utf8_chars() {
     assert!((!from_utf8(&[0xf0, 0xff, 0x10]).is_ok()));
     assert!((!from_utf8(&[0xf0, 0xff, 0xff, 0x10]).is_ok()));
 }
+
+#[test]
+fn utf8_char_counts() {
+    let strs = [("e", 1), ("é", 1), ("€", 1), ("\u{10000}", 1), ("eé€\u{10000}", 4)];
+    let mut reps = vec![1, 8, 64, 256, 512, 1024];
+    if cfg!(not(miri)) {
```

Comment:
These miri checks are just because I know it's a problem if very large test cases run in miri, because miri is slow. It should still pass there.

Reply:
The reason is there's a sense with which the loop is very unrolled (since it has the outer 192 loop and then the inner loop), so we need to be sure to use large strings to exercise the different codepaths.

Reply:
I've renamed `old_libcore` to `filter_count_cont_bytes` and `new_libcore` to `libcore`. I think it is worth keeping, since it's now identical to what we use in the small string path, and for misaligned head/tail (for now, although eventually perhaps unaligned reads for these would be better).

I don't particularly care about the other impls in benchmarks, but they're typically how you find this implemented in other code, so I've left them (but am willing to delete them if you'd like).